sw/airborne/subsystems/ins/vf_extended_float.c

```diff
@@ -178,61 +213,41 @@ void vff_propagate(float accel, float dt)
  * // update covariance
  * Pp = Pm - K*H*Pm;
  */
-static void update_baro_conf(float z_meas, float conf)
+static void update_biased_z_conf(float z_meas, float conf)
 {
+  static float K[VFF_STATE_SIZE];
+  static float P[VFF_STATE_SIZE];
+
   vff.z_meas_baro = z_meas;
 
   const float y = z_meas - vff.z + vff.offset;
   const float S = vff.P[0][0] - vff.P[0][3] - vff.P[3][0] + vff.P[3][3] + conf;
-  const float K0 = (vff.P[0][0] - vff.P[0][3]) * 1 / S;
-  const float K1 = (vff.P[1][0] - vff.P[1][3]) * 1 / S;
-  const float K2 = (vff.P[2][0] - vff.P[2][3]) * 1 / S;
-  const float K3 = (vff.P[3][0] - vff.P[3][3]) * 1 / S;
-
-  vff.z       = vff.z       + K0 * y;
-  vff.zdot    = vff.zdot    + K1 * y;
-  vff.bias    = vff.bias    + K2 * y;
-  vff.offset  = vff.offset  + K3 * y;
-
-  const float P0 = vff.P[0][0] - vff.P[3][0];
-  const float P1 = vff.P[0][1] - vff.P[3][1];
-  const float P2 = vff.P[0][2] - vff.P[3][2];
-  const float P3 = vff.P[0][3] - vff.P[3][3];
-
-  vff.P[0][0] -= K0 * P0;
-  vff.P[0][1] -= K0 * P1;
-  vff.P[0][2] -= K0 * P2;
-  vff.P[0][3] -= K0 * P3;
-  vff.P[1][0] -= K1 * P0;
-  vff.P[1][1] -= K1 * P1;
-  vff.P[1][2] -= K1 * P2;
-  vff.P[1][3] -= K1 * P3;
-  vff.P[2][0] -= K2 * P0;
-  vff.P[2][1] -= K2 * P1;
-  vff.P[2][2] -= K2 * P2;
-  vff.P[2][3] -= K2 * P3;
-  vff.P[3][0] -= K3 * P0;
-  vff.P[3][1] -= K3 * P1;
-  vff.P[3][2] -= K3 * P2;
-  vff.P[3][3] -= K3 * P3;
-
-}
 
-void vff_update_baro(float z_meas)
-{
-  update_baro_conf(z_meas, vff.r_baro);
-}
+  int i, j;
+  for (i = 0; i < VFF_STATE_SIZE; i++) {
+    K[i] = (vff.P[i][0] - vff.P[i][3]) / S;
+    P[i] = vff.P[0][i] - vff.P[3][i];
+  }
 
-void vff_update_baro_conf(float z_meas, float conf)
-{
-  if (conf < 0.f) { return; }
+  vff.z           += K[0] * y;
+  vff.zdot        += K[1] * y;
+  vff.bias        += K[2] * y;
+  vff.offset      += K[3] * y;
+  vff.obs_height  += K[4] * y;
 
-  update_baro_conf(z_meas, conf);
+  for (i = 0; i < VFF_STATE_SIZE; i++) {
+    for (j = 0; j < VFF_STATE_SIZE; j++) {
+      vff.P[i][j] -= K[i] * P[j];
+    }
+  }
+#ifdef VFF_EXTENDED_NON_FLAT_GROUND
```

Comment:
This new flag is not documented in the module xml I think.
Also, in this case, is it really`idef` or `ifndef`. This part seems to drive the obstacle height to zero. Should this be called when the feature is activated or not ? When not activated, what prevents the internal state to diverge ?

Reply:
The internal state should not be updated when the VFF_EXTENDED_NON_FLAT_GROUND is not set since vff.P[4][4] = 0.f;. As such the entire row and column of the P matrix is never updated. I confirmed this on-board.

I removed the ifdef for safety, who knows...

I also added some documentation and updated the vff message with some extra info. This is dependent on https://github.com/paparazzi/pprzlink/pull/91.